[2.9 ACRA] CardBrowser crash with sort settings

Looks like a simple type conversion issue

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/cf53b2cf-7ed7-4185-85f5-abe5defe705b

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$3.done(AsyncTask.java:325)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
at java.util.concurrent.FutureTask.run(FutureTask.java:242)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
at java.lang.Thread.run(Thread.java:762)
Caused by: java.lang.RuntimeException: org.json.JSONException: Value 0 at sortBackwards of type java.lang.Integer cannot be converted to boolean
at com.ichi2.libanki.Finder._order(Finder.java:20)
at com.ichi2.libanki.Finder._findCardsForCardBrowser(Finder.java:6)
at com.ichi2.libanki.Finder.findCardsForCardBrowser(Finder.java:1)
at com.ichi2.libanki.Collection.findCardsForCardBrowser(Collection.java:1)
at com.ichi2.async.DeckTask.doInBackgroundSearchCards(DeckTask.java:7)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:36)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:1)
at android.os.AsyncTask$2.call(AsyncTask.java:305)
at java.util.concurrent.FutureTask.run(FutureTask.java:237)
... 4 more
Caused by: org.json.JSONException: Value 0 at sortBackwards of type java.lang.Integer cannot be converted to boolean
at org.json.JSON.typeMismatch(JSON.java:100)
at org.json.JSONObject.getBoolean(JSONObject.java:413)
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
@@ -468,6 +468,10 @@ public class CardBrowser extends NavigationDrawerActivity implements
             if (mOrder == 1 && preferences.getBoolean("cardBrowserNoSorting", false)) {
                 mOrder = 0;
             }
+            //This upgrade should already have been done during
+            //setConf. However older version of AnkiDroid didn't call
+            //upgradeJSONIfNecessary during setConf, which means the
+            //conf saved may still have this bug.
             mOrderAsc = Upgrade.upgradeJSONIfNecessary(getCol(), getCol().getConf(), "sortBackwards", false);
             // default to descending for non-text fields
             if (fSortTypes[mOrder].equals("noteFld")) {

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -34,6 +34,7 @@ import com.ichi2.async.DeckTask;
 import com.ichi2.compat.CompatHelper;
 import com.ichi2.libanki.hooks.Hooks;
 import com.ichi2.libanki.template.Template;
+import com.ichi2.upgrade.Upgrade;
 import com.ichi2.utils.VersionUtils;
 
 import org.json.JSONArray;
@@ -1908,6 +1909,14 @@ public class Collection {
 
 
     public void setConf(JSONObject conf) {
+        // Anki sometimes set sortBackward to 0/1 instead of
+        // False/True. This should be repaired before setting mConf;
+        // otherwise this may save a faulty value in mConf, and if
+        // it's done just before the value is read, this may lead to
+        // bug #5523. This bug should occur only for people using anki
+        // prior to version 2.16 and has been corrected with
+        // dae/anki#347
+        Upgrade.upgradeJSONIfNecessary(this, conf, "sortBackwards", false);
         mConf = conf;
     }
 

